pkg/networkserver/redis/lua/deviceMatch.lua

```diff
@@ -0,0 +1,85 @@
+-- Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+-- ARGV[1]	- 2 LSB of FCnt (same as 16-bit FCnt field in MAC frames)
+-- ARGV[2] 	- output TTL in milliseconds
+--
+-- KEYS[1] 	- previous matching result key
+--
+-- KEYS[2] 	- sorted set of uids of devices matching current session DevAddr using 16-bit frame counters sorted by LastFCntUp
+-- KEYS[3] 	- sorted set of uids of devices matching current session DevAddr using 32-bit frame counters sorted by 2 LSB of LastFCntUp
+-- KEYS[4] 	- set of uids of devices matching pending session DevAddr
+-- KEYS[5]  - set of uids of devices matching either current or pending session DevAddr (legacy)
+--
+-- KEYS[6] 	- sorted list of uid, LastFCntUp, FNwkSIntKey, MACVersion of matching devices using 16-bit frame counters
+-- KEYS[7] 	- sorted list of uid, LastFCntUp, FNwkSIntKey, MACVersion of matching devices using 16-bit frame counters being processed
+--
+-- KEYS[8] 	- sorted list of uid, LastFCntUp, FNwkSIntKey, MACVersion of matching devices using 32-bit frame counters with no rollover
+-- KEYS[9] 	- sorted list of uid, LastFCntUp, FNwkSIntKey, MACVersion of matching devices using 32-bit frame counters with no rollover being processed
+--
+-- KEYS[10]	- sorted list of uid, LastFCntUp, FNwkSIntKey, MACVersion of matching devices using 32-bit frame counters with rollover
+-- KEYS[11] - sorted list of uid, LastFCntUp, FNwkSIntKey, MACVersion of matching devices using 32-bit frame counters with rollover being processed
+--
+-- KEYS[12] - list of uids of devices matching pending session DevAddr
+-- KEYS[13] - list of uids of devices matching pending session DevAddr being processed
+--
+-- KEYS[14] - list of uids of devices matching either current or pending session DevAddr not present in either KEYS[2], KEYS[3], nor KEYS[4]
+-- KEYS[15] - list of uids of devices matching either current or pending session DevAddr not present in either KEYS[2], KEYS[3], nor KEYS[4] being processed
+-- NOTE: It is expected that count of devices using 16-bit frame counters << count of devices using 32-bit frame counters.
```

Comment:
What does this mean exactly?

Reply:
We expect most devices to use 32-bit frame counters

Reply:
Okay, but are there any implications if that's not the case?

Reply:
Not really - the implementation is optimized for this case (which, if not now, will be true in the future as newer LoRaWAN spec versions are more widely adopted), that's it.

Reply:
Then maybe it's better to write something that says that it's optimized for devices with 32-bit counters. The way I read the current comment, it sounds like unexpected things may happen if the property doesn't hold.